.github/workflows/manual-ci.yml

```diff
@@ -0,0 +1,218 @@
+# This file is auto-generated
+# This Travis job script has NOT been generated by a script via
+#
+#   haskell-ci '--config=cabal.haskell-ci' github 'cabal.project'
+#
+# To regenerate the script (for example after adjusting tested-with) run
+#
+#   haskell-ci regenerate
+#
+# For more information, see https://github.com/haskell-CI/haskell-ci
+#
+# REGENDATA ["--config=cabal.haskell-ci","github","cabal.project"]
+name: Manual experiment
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+
+jobs:
+  linux:
+    name: Haskell-CI Linux
+    strategy:
+      matrix:
+        include:
+        - image: phadej/ghc:8.10.1-bionic
```

Comment:
The use of Docker here is interesting. Is there an advantage to doing this over something like [`setup-haskell`](https://github.com/actions/setup-haskell)?

Reply:
I'm just experimenting here. I won't use `setup-haskell` for linux builds, it doesn't really add anything (maybe marginal speed-up?)

Reply:
The real case for using docker, is that with `docker` one gets much more cleaner environment, where the GitHub builder image is full of stuff and changing all the time.

Reply:
Gotcha. I was only asking since I assumed that `setup-haskell` simplified the process of actually installing `ghc` and `cabal`. But if you can get GHA to do the same thing with manual invocations of `apt-get install`, then that's fine too.

Reply:
It would "simplify" for macOS where one don't have anything like Docker so (unfortunately) have to run on GitHub provided image, then `setup-haskell` would be able to use pre-installed GHCs.

Note, that installing with `apt` takes about 30 seconds, i.e. barely anything in larger scheme of things. On macOS `ghcup` uses bindists which `configure` and `make install` take noticeable time. So there it matters.

Reply:
haskell/actions/setup (formerly setup-haskell) takes care of a few things:

1. Cross platform support with the same syntax. This is actually what I consider to be the most important aspect of it. Specifying "I want this version of GHC and this version of cabal-install on my path" with the same syntax regardless of the underlying OS is very powerful.
2. Caching, speed, performance. This is all really nice stuff, but not necessarily essential; if you already have the right version of GHC+cabal-install setup in your PATH, the action will no-op. This allows you to do things like use `haskell/actions/setup` multiple times in a single job and pay the most minimal cost possible. It also makes using it a zero cost abstraction for self hosted runners. To put it another way, I go to pretty far lengths to make sure there's "no reason" to avoid using it.
3. Integration with github actions. This is another "real"/main benefit of using it. As it always runs natively, you don't have to worry about docker/filesystem abstractions leaking. It also takes care to not do anything "unexpected", so it will always compose well with other github actions; it is much easier to chain `haskell/actions/setup` with something like `actions/create-release`, for example.
   It also sets convenience outputs so that you can reference, eg, the cabal store (or stack root) in a platform-agnostic way for setting up caching easier or whatever you want.
   Going further, `haskell/actions/setup` could integrate things like options to automatically cache dist-newstyle using something like `cabal-cache`, or setting up "matchers" so that GHC/cabal/etc's errors or warnings had tighter integration with github action's UI.

As for what that looks like, implementation wise: for linux, it checks to see if GHC and/or cabal-install was already installed. If not, it'll try hvr's PPA and use that. If that doesn't work, it'll download `ghcup` and use that to install things. If stack is enabled, it'll do the same thing for stack (check if installed and download from github releases if needed). Lastly, it sets the PATH, and outputs a few convenience things like the location of cabal's store or the stack root.

So, yes, one can easily do this on linux themselves by just blindly installing GHC using hvr's ppa and the setting the path themselves. It will take slightly more time for pre-installed GHC's than otherwise because apt-get will be invoked rather than just setting the path. On the other hand, that time might be washed out by github actions actually cloning this repo and then running the right node program. So... meh; it's kind of a wash if you don't care about any of the other things `haskell/actions/setup` offers.

In the end, whether all of that *matters* for `haskell-ci` is an entirely different question. If `haskell-ci` wants to enable cross platform testing conveniently, it will eventually end up re-implementing the majority of `haskell/actions/setup`'s logic, so it will make sense to integrate other tooling where possible (from a sheer effort perspective), but that might not take the form of actually using `haskell/actions/setup`, particularly if the interest is keeping the exact same logic across different CI platforms + locally.

Reply:
> Lastly, it sets the PATH

By the way, that is a thing which `haskell-ci` **does not** do. There is no `cabal` or `ghc` in `PATH`. That breaks some "fihsy" stuff where authors depends on compiler to be `ghc` and in PATH. (They could use `ghc-paths` which at least explicitly declares that kind of nastyness).

Reply:
> By the way, that is a thing which `haskell-ci` **does no**t do.

Yeah, it's interesting. That approach works well when you have the entire haskell testing setup self contained, but the way github actions tend to be structured is by starting with a clean environment, mutating the environment to setup, and then running the tests. So the action needs to add relevant programs to the PATH to integrate well. I do also export paths for the directory + binary for each of ghc, stack, and cabal-install as outputs, but I have never actually seen them used.

Interestingly, I can't actually wipe the path inside an action. The design of `echo FOO=bar >> $GITHUB_PATH` prevents this (intentionally). So I "have" to do that design. `haskell-ci` could, of course, generate a script where the first part of every step resets/wipes the PATH of all haskell related paths.